data/src/test/java/org/apache/iceberg/RecordWrapperTest.java

```diff
@@ -0,0 +1,120 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+import org.apache.iceberg.types.Types;
+import org.apache.iceberg.util.StructLikeWrapper;
+import org.junit.Assert;
+import org.junit.Test;
+
+import static org.apache.iceberg.types.Types.NestedField.optional;
+import static org.apache.iceberg.types.Types.NestedField.required;
+
+public abstract class RecordWrapperTest {
+
+  private static final Types.StructType PRIMITIVE_WITHOUT_TIME = Types.StructType.of(
+      required(100, "id", Types.LongType.get()),
+      optional(101, "data", Types.StringType.get()),
+      required(102, "b", Types.BooleanType.get()),
+      optional(103, "i", Types.IntegerType.get()),
+      required(104, "l", Types.LongType.get()),
+      optional(105, "f", Types.FloatType.get()),
+      required(106, "d", Types.DoubleType.get()),
+      optional(107, "date", Types.DateType.get()),
+      required(108, "ts_tz", Types.TimestampType.withZone()),
+      required(110, "s", Types.StringType.get()),
+      required(112, "fixed", Types.FixedType.ofLength(7)),
+      optional(113, "bytes", Types.BinaryType.get()),
+      required(114, "dec_9_0", Types.DecimalType.of(9, 0)),
+      required(115, "dec_11_2", Types.DecimalType.of(11, 2)),
+      required(116, "dec_38_10", Types.DecimalType.of(38, 10))// maximum precision
+  );
+
+  private static final Types.StructType TIMESTAMP_WITHOUT_ZONE = Types.StructType.of(
+      required(101, "ts0", Types.TimestampType.withoutZone()),
+      required(102, "ts1", Types.TimestampType.withoutZone())
+  );
+
+  private static final Types.StructType TIME = Types.StructType.of(
+      required(100, "time0", Types.TimeType.get()),
+      optional(101, "time1", Types.TimeType.get())
+  );
+
+  @Test
+  public void testSimpleStructWithoutTime() {
+    generateAndValidate(new Schema(PRIMITIVE_WITHOUT_TIME.fields()));
+  }
+
+  @Test
+  public void testTimestampWithoutZone() {
+    generateAndValidate(new Schema(TIMESTAMP_WITHOUT_ZONE.fields()));
+  }
+
+  @Test
+  public void testTime() {
+    generateAndValidate(new Schema(TIME.fields()), (message, expectedWrapper, actualWrapper) -> {
+      for (int pos = 0; pos < TIME.fields().size(); pos++) {
+        Object expected = expectedWrapper.get().get(pos, Object.class);
+        Object actual = actualWrapper.get().get(pos, Object.class);
+        if (expected == actual) {
+          return;
+        }
+
+        if (expected == null || actual == null) {
+          Assert.fail(String.format("The expected value is %s but actual value is %s", expected, actual));
+        }
+
+        int expectedMilliseconds = (int) ((long) expected / 1000_000);
+        int actualMilliseconds = (int) ((long) actual / 1000_000);
+        Assert.assertEquals(message, expectedMilliseconds, actualMilliseconds);
```

Comment:
I don't think that this is valid. Millisecond truncation should never be allowed, except in Flink where it is unavoidable. In the data module, this is incorrect.

Reply:
You are correct,  my. original idea was to put this method in the flink test class , but forget to move it to...